### DISCUSSION TITLE
Is there a smart method to 'blur' sealant stripes and wall ceiling transitions

### DISCUSSION THREAD

**Discussion by VivitionDeveloper** [libvips/libvips#4155] [Q&A] [answered]
Consider the following photo (intentionally taken a white car):

![image](https://github.com/user-attachments/assets/58830909-03c7-46db-8968-0a020b818d25)

Is there a method in libvips to 'blur' the sealand stripes (marked in yellow) or wall/ceiling transitions (marked red)? 
![image](https://github.com/user-attachments/assets/4619989f-03a9-4775-a18a-2258848d9f2b)

Of course, without removing information from the car. Actually, these blur actions are only wanted in the 'upper half' of the photo. I put 'upper' in quotes, because 'upper' depends on the height of the car whereof the picture was taken (a car can be higher or lower).

The goal is either to blur visiable sealant stripes or wall/ceiling crossings or give them a smooth (near white, or some 'average' between the whites of the wall and the greys of the ceiling) color with or without gradient. Any suggestions are welcome.

**Comment by jcupitt** [comment]
Hello again,

Sure, you can blur the whole image, then use a mask to pick pixels from the blurred version or the unblurred. You can soften the mask and get feathered edges too, which helps. Because libvips calculates pixels on demand, it won't actually blur the whole background, just the bits it needs for the final image.

The tricky part of course is making a mask. Do you need this to be automatic, or can someone draw it? If your camera doesn't move much, I suppose a hand-drawn mask would be OK.

**Reply by VivitionDeveloper** [reply]
I need this automatic. And we have a possibility to calculate/find the top roof coordinates (i.e. bounding box) of the car. Great answer again. I'll ask my boss to send some gift to you ;-)

**Reply by jcupitt** [reply]
Heh ok, thanks! The "donate" button at the top is the simplest.

**Reply by VivitionDeveloper** [reply]
I think the automatic part is the most difficult part. If you have good ideas, you are welcome.

**Reply by VivitionDeveloper** [reply]
> Heh ok, thanks! The "donate" button at the top is the simplest.

I will ask my boss, but did not see the donate button...

**Reply by jcupitt** [reply]
It's called "Sponsor" and has a red heart on. You might need to go to the project page on github, they don't always show it on subpages.

**Comment by jcupitt** [comment]
I had a quick go in nip2:

![image](https://github.com/user-attachments/assets/f3a7db1d-98ab-43e7-93b0-2db9f9379ba7)

I made the worst mask possible, then used that to blend smoothly between your car photo and a very blurred version of the car. It makes:

![car2](https://github.com/user-attachments/assets/7612d4d3-3591-4358-a8f5-239a5964eedf)

You can still see the top triangle as a shadow, I don't know if that's a problem.

**Reply by VivitionDeveloper** [reply]
If you have some pointers how to achieve this in C/C++, it would be very helpful. The result looks promising.

Abnd how did you determine the gradient? I.e. how can we do that in an automated way?

Also, is there a safe way to 'blur' the vertical sealant stripes in the wall, without a big risk in unwanted blur of the car?

**Comment by jcupitt** [comment]
Sorry, which gradient? I just drew the black area with my mouse.

Yes, automatic mask generation would be the hard part, it would need a fair amount of work. 

You can see all the steps in the screenshot, or I put the nip2 workspace here: 

www.rollthepotato.net/~john/car.ws

**Reply by VivitionDeveloper** [reply]
My mistake. I assumed that you have used a gradient to blur the ceiling (of the studio). But you have used a blur. However, I'm playing with the idea to use a 'white' gradient to mask the ceiling. Not sure whether that is a feasible or useful idea.

When I open the screenshot for enlarging it, I got a github error, so it was quite hard to read what was done. Thanks for providing the nip2 workspace.

**Comment by VivitionDeveloper** [comment]
I finally have time to pickup this issue again. Some questions popped up:

- I discovered that nip2 apparentl y uses image magick for gaussian blur, is this correct? 
- When this is correct: how can I move to similar results in libvips with using vips_gaussblur(), i.e. what are the vips counterpart values for min_ampl (0.1?), precision (I guess VIPS_PRECISION_INTEGER)?
- What vips operation is performed in cell A5, i.e. what is the meaning of the ++ operator? Is this just the same as operator+= as listed on https://www.libvips.org/API/current/cpp/classVImage.html?

And a little bit different questions:
What is expected to give better performance and results for the background:

1. downscale input image + gaussian blur + upscale image and then apply the background as overlay to the mask?
2. or use full ress inout image and apply gaussian blur?
A quick test learned that 1 is probably faster, but I cannot easily see differences in output quality.

An finally an answer to something that I forgot to answer before:

> You can still see the top triangle as a shadow, I don't know if that's a problem.

The triangle is indeed a problem and becomes more problematic when a larger part of the ceiling is visible (depends on camera zoom level and the size of the studio). I think that increasing the sigma for the gaussian blur will help, but also increases the risk that the roof of a car is hit (i.e. gets unwanted blur applied). E.g. when a car is higher (mini VAN/bus) and the camera position is lower, then this risk increases. If you have a suggestion to eliminate this risk, your suggestions are welcome.

**Reply by jcupitt** [reply]
> I discovered that nip2 apparentl y uses image magick for gaussian blur, is this correct?

No, it's in libvips. There's a "magick" menu with a lot of basic IM operations in, perhaps you found that? They are a lot slower than the libvips equivalents.

`++` is join, so `[1, 2, 3] ++ [4, 5, 6] == [1, 2, 3, 4, 5, 6]`. Applied to images it does bandwise-join, so `R ++ G ++ B` makes an RGB image.

> The triangle is indeed a problem and becomes more problematic when a larger part of the ceiling is visible

Then we should probably infill that part with white. I used a blur because it was slightly simpler, but white would be pretty easy too.

**Reply by VivitionDeveloper** [reply]
Wow, thanks for quick response. 

> No, it's in libvips. There's a "magick" menu with a lot of basic IM operations in, perhaps you found that? They are a lot slower than the libvips equivalents.

The reason why I thought this, is that vips8 does not have a vips_gaussblur (vips8) that uses radius, or mask shape. May be this is vips7 related? Indirectly, vips_gaussblur uses vips_gausmat, which has a 'precision' option (and in your NIP2 example, integer precision was selected). That's why next question popped up:

Does seting the precision to integer have any impact?

> ++ is join, so [1, 2, 3] ++ [4, 5, 6] == [1, 2, 3, 4, 5, 6]. Applied to images it does bandwise-join, so R ++ G ++ B makes an RGB image.

As cell A4 is a single band alpha, is then the join similar to a bandjoin of the first 3 channels (RGB) from the input image (=A1) and the blurred alpha (4th channel) from A4? I.e. in pseudo code in vips: when A1 is a RGB VipsImage img, then the belonging operation is img.bandjoin(cell A4)?

Do you have suggestions for min_ampl values, precision and probably hints for 'assumed' better performance (i.e. with or without downscaling first)?

**Reply by jcupitt** [reply]
Ah OK, `sigma` is the same as `radius`, but more commonly used in image processing. You can see the mask it uses with eg.:

```
$ VIPS_INFO=1 vips gaussblur k2.jpg x.jpg 5
VIPS-INFO: 11:39:09.960: VIPSHOME = /home/john/vips
VIPS-INFO: 11:39:09.960: VIPS_PREFIX = /home/john/vips
VIPS-INFO: 11:39:09.960: VIPS_LIBDIR = /home/john/vips/lib
VIPS-INFO: 11:39:09.960: prefix = /home/john/vips
VIPS-INFO: 11:39:09.960: libdir = /home/john/vips/lib
VIPS-INFO: 11:39:09.968: gaussblur mask width 17
VIPS-INFO: 11:39:09.968: threadpool completed with 2 workers
VIPS-INFO: 11:39:09.968: threadpool completed with 2 workers
VIPS-INFO: 11:39:09.968: threadpool completed with 2 workers
VIPS-INFO: 11:39:09.969: threadpool completed with 2 workers
VIPS-INFO: 11:39:09.969: convi: using vector path
VIPS-INFO: 11:39:09.969: threadpool completed with 1 workers
VIPS-INFO: 11:39:09.969: threadpool completed with 1 workers
VIPS-INFO: 11:39:09.969: threadpool completed with 1 workers
VIPS-INFO: 11:39:09.969: convi: using vector path
VIPS-INFO: 11:39:10.026: threadpool completed with 1 workers
vips_image_dispose: temp-16 vips_convi_uchar_vector_gen computed 111%
```

So sigma 5 makes a mask 17 points across, or radius 8.

Integer precision is a lot faster, but less accurate. Though the inaccuracy is pretty small and you'll probably only care if you are doing scientific work.

```
john@banana ~/pics $ time vips gaussblur nina.jpg x.jpg 20 --precision integer

real	0m0.318s
user	0m0.943s
sys	0m0.140s
john@banana ~/pics $ time vips gaussblur nina.jpg x.jpg 20 --precision float

real	0m0.577s
user	0m13.612s
sys	0m0.210s
```

int is more than 10x faster.

**Reply by jcupitt** [reply]
Yes, `++` is bandjoin on images. You can join constants and arrays to images too.

I think I'd leave `min_ampl` and `precision` at the default settings.

**Reply by VivitionDeveloper** [reply]
Thanks again. I will try to convert the nip2 workspace to a C++ libvips related one and see whether the outcome is almost the same.

Just one question left: what is the rationale behind using a larger radius (or sigma) for B3 vs A4 in the NIP2 example?

And when is NIP4 available (on Windows), i.e. that's probably more close to the vips8 interface (where NIP2 uses other input values).

**Reply by jcupitt** [reply]
I don't think there's a reason, it was just where I dragged the slider. Any large value should be fine.

nip4 is still a work in progress. I'd guess six to nine months, but who knows!

**Reply by VivitionDeveloper** [reply]
Good luck with nip4. Hopefully you will reach that milestone in the mentone time frame. 

I was able to write something in C++ that produces similar results as in NIP2 (omitting layers, fac, shape, that are not available with vips_gaussblur). For now, I will rely on the libvips defaults here. I will try to play around with best balance between result, performance, sigma etc. My goal (or mission) is to 'blur the ceiling' in less than 2s for a 6000x4000 image. That's somewhat a challenging job. And maybe filling the triangle with white (or a picked up color, close to white from the wall) is indeed required to remove the triangle.

**Reply by VivitionDeveloper** [reply]
I just ran into a crash with a VImage::option() that was reused in a seconds VIPS gausblurr() (C++) call. Apparently, a call to gaussblur changes vOptions. I.e. options is no const, but changed (I did not expect that). See the following piece of (partly code from us, using an own wrapper class):

```
constexpr double dGaussBlurr = 50;
auto imgMask = viMask.getVipsImage();
auto imgAlpha = imgMask[3];
auto* vOptions = vips::VImage::option();
vOptions->set("min_ampl", 0.1);
vOptions->set("precision", VipsPrecision::VIPS_PRECISION_INTEGER);
imgAlpha = imgAlpha.gaussblur(dGaussBlurr, vOptions);
```
After the call to imgAlpha.gaussblur, vOptions changed and the internal vector its size becomes different. Not sure whether this is defined behavior or something wrong. The used vips version is 8.15.3 (static libs), but 8.16 runtime (I'm underway to fully move to 8.16).

When another VIPS (gaussblur) call reuses this vOptions, then the program will croak (of course).

**Reply by jcupitt** [reply]
You need a new option for every call.

**Comment by VivitionDeveloper** [comment]
When a more white triangle is needed, what suggestions do you have? Add a white (or picked color white from the wall or on top left or top right ceiling), fill a block somewhere in de top middle ceiling part and rely on the blur for more whites? Or something else? Your suggestion is welcome.

**Comment by VivitionDeveloper** [comment]
I've tried some things with NIP2 again. When filling the triangle wtih pure white (RGB 255, 255, 255, similar to the value in the edges) and applying that mask to the background, then the final result looks ugly, see picture. Do you have a solution for this?
![image](https://github.com/user-attachments/assets/6920e884-be38-4b8d-909f-d965017ea4ab)

Even when using a not-perfect-white, but near white (RGB 242, 242, 244), then the result is not hmm, wonderful.
![image](https://github.com/user-attachments/assets/e19de45c-8bae-423a-ba19-a78785b19e33)

I don't know how to fix this.

BTW. I just saw that this post in the Help wanted category, not sure whether this is the right category.

**Reply by VivitionDeveloper** [reply]
Can anyone provide an answer to previous question about 'how to fill the mask' with a (for a human) pleasant result, without hard (visible) transitions?

**Comment by jcupitt** [comment, marked as answer]
I found time to have a go with another approach -- using `fill_nearest`:

https://www.libvips.org/API/current/libvips-morphology.html#vips-fill-nearest

This fills in any zero pixels with the nearest non-zero pixel. Here's a nip2 workspace:

http://www.rollthepotato.net/~john/car2.ws

It looks like this:

![image](https://github.com/user-attachments/assets/12b58b99-75b2-4462-844f-61f4943ccc6e)

Doubleclick on H1, then drag H10 up and down to mark the low point of the ceiling.

I found the ceiling was mostly missing blue, so slider H11 sets a ceiling threshold -- a value for blue below which we think that pixel should be replaced.

The mask is then just (blue < H11 AND y < H10). I added a small blur to move the mask outwards by 10 pixels, then set those pixels black and used `fill_nearest` to replace them.

`fill_nearest` is a vips8 thing, so it's not in nip2. I ran it with a `vips_call`.

Here's the result:

![car2](https://github.com/user-attachments/assets/b3d768e1-643d-46f9-ac7c-9c1f532c1f4a)

If you boost the contrast a lot you can see what it's done:

![image](https://github.com/user-attachments/assets/e2b0dc51-9d10-4917-a95f-9e9811f7157c)

So pixels below the triangle have been pulled up to fill it.

**Comment by VivitionDeveloper** [comment]
Thanks again. I will take a look on it and try the things you did (but in C++).

**Comment by VivitionDeveloper** [comment]
Just for reference or useful for other readers: the blur method worked pretty well, but requires manual selection of the area to blur. And when the ceiling becomes bigger, then nice blurring becomes more difficult, i.e. remains visible to some extent. Another issue popped up, i.e. when the camera position is lower, the car to photograph bigger/higher and the studio smaller, then the roof of the car 'hits' the ceiling, making a simple blur impossible (as then the car is unintentionally blurred too).

The performance of blurring a full-res (6000x4000px) photo took almost 2s (with a manually created mask and downscaling + upscaling the background) on our hardware (server grade AMD Epyc with much cores). The used sigma for the blur was 30 (on the original input + also on the 10x downscaled background) and increasing or lowering the sigma has a big effect on performance.
